js/src/main/scala/scalajson/ast/JValue.scala

```diff
@@ -77,22 +77,26 @@ object JNumber {
   def apply(value: Float): JValue = value match {
     case n if java.lang.Float.isNaN(n) => JNull
     case n if n.isInfinity => JNull
-    case _ => JNumber(value.toString)
+    case _ => new JNumber(value.toString)
   }
+
+  def fromString(value: String): Option[JNumber] =
+    if (value.matches(jNumberRegex))
+      Some(new JNumber(value))
+    else
+      None
+
+  def unapply(arg: JNumber): Option[String] = Some(arg.underlying)
```

Comment:
The `unapply` for JNumber is there for these reasons

1. Consistency in the AST (everything is a case class, containing a single `value` with an unapply)
2. The string can be directly converted to the number you want (i.e. `toDouble`, `toInt`) since its always going to be a valid number. This may be improved with https://github.com/mdedetrich/scalajson/issues/17#issuecomment-312158304